aerogear/keycloak-operator
# Provision keycloak once a keycloak custom resource is created

## Description
Implement the provision of a Keycloak instance based on the definition of the Keycloak custom resource. A Keycloak instance and an admin user for that instance should be created. A reference to the admin credential is added to the Keycloak custom resource as well as the service instance id.

## Progress
- [x] Provision Keycloak when a Keycloak custom resource is created.
- [x] Create a secret in the same namespace for the admin credentials and add a reference of this to the Keycloak CR.
- [x] Add a reference of the Keycloak service instance to the Keycloak CR.
- [x] Set phase to `complete` and status ready to `true` once Keycloak has successfully provisioned.

## Additional Notes
Related JIRA Ticket - https://issues.jboss.org/browse/AEROGEAR-7683

**Verification Steps**
1. Create a [Keycloak custom resource](https://github.com/aerogear/keycloak-operator/blob/master/deploy/examples/keycloak_min.json) in the same namespace that the operator is deployed in.
2. Keycloak should start provisioning in that namespace.
3. A secret `keycloak-admin-cred-<hash>` should be created within the same namespace. This should have the following properties: 
```
ADMIN_USERNAME: admin
ADMIN_PASSWORD: <generated-password>
```
4. The Keycloak custom resource should have `status.phase: provisioning` and  the following properties added to its `Spec`: 
```
Spec:
   adminCredentials: <keycloak-admin-cred-ref>
   instanceID: <keycloak-service-instance-id>
   ...
```
5. Once provision is complete, the Keycloak custom resource should have `status.phase: complete`
6. Verify that you can login to the Keycloak administration console with the credentials provided in the `keycloak-admin-cred-<hash>` secret.
